moshi/src/main/java/com/squareup/moshi/AdapterMethodsFactory.kt

```diff
@@ -0,0 +1,362 @@
+                moshi.nextAdapter(factory, parameterTypes[0], qualifierAnnotations)
+              } else {
+                moshi.adapter(parameterTypes[0], qualifierAnnotations)
+              }
+            }
+
+            override fun fromJson(moshi: Moshi, reader: JsonReader): Any? {
+              val intermediate = delegate.fromJson(reader)
+              return invoke(intermediate)
+            }
+          }
+        }
+        else -> {
+          throw IllegalArgumentException(
+            "Unexpected signature for $method.\n@FromJson method signatures may have one of the following structures:\n    <any access modifier> R fromJson(JsonReader jsonReader) throws <any>;\n    <any access modifier> R fromJson(JsonReader jsonReader, JsonAdapter<any> delegate, <any more delegates>) throws <any>;\n    <any access modifier> R fromJson(T value) throws <any>;\n"
+          )
+        }
+      }
+    }
+
+    /** Returns the matching adapter method from the list. */
+    private fun get(
+      adapterMethods: List<AdapterMethod>,
+      type: Type,
+      annotations: Set<Annotation>
+    ): AdapterMethod? {
+      for (adapterMethod in adapterMethods) {
+        if (Types.equals(adapterMethod.type, type) && adapterMethod.annotations == annotations) {
+          return adapterMethod
+        }
+      }
+      return null
+    }
+  }
+
+  internal abstract class AdapterMethod(
+    private val adaptersOffset: Int,
+    type: Type,
+    parameterCount: Int,
+    val annotations: Set<Annotation>,
+    val adapter: Any,
+    val method: Method,
+    val nullable: Boolean
+  ) {
+    val type = type.canonicalize()
+    private val jsonAdapters: Array<JsonAdapter<*>?> = arrayOfNulls(parameterCount - adaptersOffset)
+
+    open fun bind(moshi: Moshi, factory: JsonAdapter.Factory) {
+      if (jsonAdapters.isNotEmpty()) {
+        val parameterTypes = method.genericParameterTypes
+        val parameterAnnotations = method.parameterAnnotations
+        for (i in adaptersOffset until parameterTypes.size) {
+          val type = (parameterTypes[i] as ParameterizedType).actualTypeArguments[0]
+          val jsonAnnotations = parameterAnnotations[i].jsonAnnotations
+          jsonAdapters[i - adaptersOffset] =
+            if (Types.equals(this.type, type) && annotations == jsonAnnotations) {
+              moshi.nextAdapter(
+                factory, type, jsonAnnotations
+              )
+            } else {
+              moshi.adapter<Any>(type, jsonAnnotations)
+            }
+        }
+      }
+    }
+
+    open fun toJson(moshi: Moshi, writer: JsonWriter, value: Any?): Unit = throw AssertionError()
+
+    open fun fromJson(moshi: Moshi, reader: JsonReader): Any? = throw AssertionError()
+
+    /** Invoke the method with one fixed argument, plus any number of JSON adapter arguments. */
+    protected fun invoke(fixedArgument1: Any?): Any? {
```

Comment:
Let's fix this parameter name to be more descriptive

Reply:
https://github.com/square/moshi/pull/1467/commits/8e9daf9e3d8e5b9ab2e0bee794a8a8434f45aef1